### PR TITLE
when aggregating non-day periods, only store one datatable and blob row in memory at a time

### LINKED ISSUE #18295: When aggregating data tables for a week, month, year or range period, don't store all archive content in memory
This is a follow up from https://github.com/matomo-org/matomo/pull/17817 where we only expand one data table at a time.

We've noticed though we're still seeing sometimes memory issues for eg a yearly archive where it exceeds 8GB. This archive has 9 levels down of subtables and overall a lot of rows. One thing that could help with memory be in **to not load the data of all data tables at once into memory**.:

image

Each data table is then later expanded one at a time. 

The only problem with loading the needed data tables on demand would be performance as we'd need to do a lot more select queries to get each "subchildren" data table. Maybe that means we might not want to do this or maybe we can find a compromise somehow.

refs L3-126

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

Fixes #18295 

Changes:
* Add Archive::querySingleBlob() method that queries archive data for a single blob using a cursor, instead of `Db::fetchAll()`. The result is ordered so the datatables can be aggregated w/o needing the entire datatable tree in memory.
* In ArchiveProcessor::aggregateDataTableRecords(), provide a default sort column if none is specified. If the table includes nb_visits, we sort by that before truncation. This ensures predictable results for blob report limiting.
* In ArchiveProcessor, use `Archive::querySingleBlob()` to query blob data and aggregate the result w/o loading entire trees in memory.
* Remove DataCollection::forEachBlobExpanded() since it is no longer used.

Notes:
* The new method of querying archive data is only used for aggregating data. This is to avoid unintended performance regressions when just querying data from the API. Additionally, using a cursor would only be more memory efficient when aggregating datatables together, not when just querying them.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Review by tsteur — COMMENTED** [review]
FYI haven't fully reviewed it yet but did some profiling.

Here's the result for this PR . The peak memory was 163MB

image
image

And here the result for the `4.x-dev` branch where it used 347MB peak.

image
image

It also executes slightly faster with this PR 👍 🎉 
This was for a date range of `2022-09-06,2022-11-26`

I can't test it yet over a longer period but I think that would be a similar result.

> The new method of querying archive data is only used for aggregating data. This is to avoid unintended performance regressions when just querying data from the API. Additionally, using a cursor would only be more memory efficient when aggregating datatables together, not when just querying them.

Looking at the test result there, maybe it could also be used in the API? That's where we're currently having problems. Or was the new implementation slower for you?

**Comment by diosmosis** [comment]
@tsteur 

> Looking at the test result there, maybe it could also be used in the API? That's where we're currently having problems. Or was the new implementation slower for you?

I didn't actually test it, but it shouldn't make a difference, it should only make a difference when aggregating (browser triggered or core:archive triggered). In this case, we fetch datatables for multiple dates and combine them, so the peak memory use can be lowered to one datatable tree. For just querying through the API, though, all the data is necessary to be returned, so it should make no difference if we use a cursor or fetch them all at once. Everything has to be in memory no matter what (unless we modify matomo renderers to stream data out instead of writing all at once, which would be an enormous change). Anyway, the SELECT query is far more complicated now and I don't know what that would do under load so I thought it was better to constrain the effects of this change for now.

**Comment by tsteur** [comment]
@diosmosis can you maybe send a link where this different behaviour is implemented? Looking at the PR I can't find different behaviour for API vs others. Or maybe I don't fully understand what is meant by API and how we need to fetch all records there anyway.

**Comment by diosmosis** [comment]
@tsteur https://github.com/matomo-org/matomo/blob/46679d1300bfc8b7f551afb2b860ebee7081ed7f/core/DataAccess/ArchiveSelector.php#L142-L244 (querying for archives to just return them) vs. https://github.com/matomo-org/matomo/blob/46679d1300bfc8b7f551afb2b860ebee7081ed7f/core/DataAccess/ArchiveSelector.php#L452-L519 (querying for a single blob to aggregate them). The difference in the generated SQL is determined by https://github.com/matomo-org/matomo/blob/46679d1300bfc8b7f551afb2b860ebee7081ed7f/core/DataAccess/ArchiveSelector.php#L559-L565

**Comment by tsteur** [comment]
Hi @diosmosis . I understand now. Debugged the code and seeing now what's happening and understanding the extra queries. Also understanding the many more queries you mean. I was actually hoping there could be only one query but then in the generator itself we only call `$cursor->fetch()` to get the next data result. I don't actually know if that would reduce memory or if in the background PHP would still hold everything in memory. Just thinking that this could improve the speed and reduce the amount of DB queries quite signficantly as these queries can get slow in production (although we improved their speed recently with a different index).

image

**Comment by diosmosis** [comment]
@tsteur 

> I was actually hoping there could be only one query but then in the generator itself we only call $cursor->fetch() to get the next data result. 

This should be what happens? There should only be one query per archive table. Can you elaborate?

EDIT: when I said 'under load' I meant 'in production when used for every API blob query, not just when aggregating' and I was referring to the complicated ORDER BY thats used to order by idSubtable. Which isn't actually needed for API requests that just fetch data without doing any aggregation.

**Comment by tsteur** [comment]
@diosmosis I'm so sorry. It actually does work like that indeed. It only looked that way as for one of the date ranges it had no data and then it would always start again at the top of the function and issue the query again but this is actually not the case when there are matching rows. I completely missed it. From my perspective this looks all good 💯 

Not sure if @matomo-org/core-reviewers  have any other thoughts otherwise? Ideally, we could even merge this into a 4.14.0 release as it would be great to have this change on the Cloud already earlier as I believe there isn't a breaking change.

**Comment by diosmosis** [comment]
> It only looked that way as for one of the date ranges it had no data and then it would always start again at the top of the function and issue the query again but this is actually not the case when there are matching rows.

Interesting, I think that probably shouldn't happen, I'll take a look there.

EDIT: Oh I see, nevermind I thought you meant the loop would go on forever in that case, but it shouldn't :)

**Comment by tsteur** [comment]
@diosmosis it doesn't. It's all good. It was just querying it for different record names which be expected. I didn't realise the record names changed while debugging 👍

File: tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml [hunk 1]
```diff
@@ -12,25 +12,25 @@
...
-		<max_actions>2</max_actions>
-		<sum_visit_length>10</sum_visit_length>
-		<bounce_count>15</bounce_count>
+		<label>Windows / Firefox / 1024x768</label>
+		<nb_visits>25</nb_visits>
+		<nb_actions>25</nb_actions>
+		<max_actions>1</max_actions>
+		<sum_visit_length>0</sum_visit_length>
+		<bounce_count>25</bounce_count>
 		<nb_visits_converted>0</nb_visits_converted>
```
Comment (sgiehl) [inline tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml:21]: Any specific reason why that test result changed? To me it looks incorrect, as there hadn't been any changes to the fixture. If the change is valid, I'm actually wondering if the changes in this PR have fixed something we weren't aware of 🤔

File: core/ArchiveProcessor.php [hunk 2]
```diff
@@ -9,11 +9,11 @@
 namespace Piwik;
 
 use Exception;
-use Piwik\Archive\DataCollection;
 use Piwik\Archive\DataTableFactory;
 use Piwik\ArchiveProcessor\Parameters;
 use Piwik\ArchiveProcessor\Rules;
 use Piwik\Container\StaticContainer;
+use Piwik\DataAccess\ArchiveSelector;
```
Comment (sgiehl) [inline core/ArchiveProcessor.php]: Unused use statement here:
```suggestion
```

**Review by sgiehl — CHANGES_REQUESTED** [review]
Had a look through the code. To me the changes are looking reasonable.
@tsteur I didn't debug the code or checked any memory usage yet, as you already did that. Or shall I invest some additional time to do that as well?

File: tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml (hunk 1 above)
Comment (diosmosis) [reply tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml:21]: The change is random, the record is not sorted before truncation (see https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Resolution/Archiver.php#L47), so the order in which blobs are aggregated determines what rows show up in this test result. Before blobs were queried w/o an ORDER BY, now they are sorted by idsubtable.

Comment (diosmosis) [reply tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml:21]: Actually, thats not entirely correct, apparently I added a default sort column before truncating so this test result wouldnt randomly change. The change is listed in the PR description.

**Comment by tsteur** [comment]
> @tsteur I didn't debug the code or checked any memory usage yet, as you already did that. Or shall I invest some additional time to do that as well?

I did check memory usage and it definitely improved and went down and also debugged the code etc.

File: tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml (hunk 1 above)
Comment (sgiehl) [reply tests/PHPUnit/System/expected/test_reportLimiting__Resolution.getConfiguration_month.xml:21]: Ah ok. That makes sense. So the result is now actually "more correct", as before it truncated records that actually had more visits.

**Review by sgiehl — APPROVED** [review]

**Pull request merged**